crypto/ffc/ffc.c

```diff
@@ -0,0 +1,1435 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/*
+ * Finite Field cryptography (FFC) is used for DSA and DH.
+ * This file contains methods for generation and validation of
+ * FFC Domain Parameters, and some helper functions for generation and
+ * validation of FFC keys.
+ *
+ * The implementations are defined in FIPS 186-4 Appendix A (for use with DSA).
+ * SP800-56Ar3 also refers to using FFC (for DH key agreement).
+ * The DH case is only used for backwards compatibility (approved safe primes
+ * should be used for new applications).
+ *
+ * NOTES:
+ *   The DSA and DH standards are complex. i.e.
+ *   The FIPS186-4 DSA/FFC Domain Parameters require p,q,seed and counter to
+ *   validate. However the ANS1 for DSA consists of just p,q,g.
+ *   The ASN1 for DH is either
+ *     (1) p,g,[qlength] (which is completely unverifiable)
+ *     (2) p,q,g, [j] [seed counter] (which is only verifiable if the optional
+ *     [seed counter] is specified]. ((2) is defined in X9.42).
+ *
+ *  Note that the gindex used for canonical g verification is not saved in ASN1.
+ *  This means that in order to verify FFC params correctly - values for
+ *  seed, counter and gindex need to be stored in order to verify p,q,g.
+ *
+ * Currently in OPENSSL dsa param generation calculates the seed and counter
+ * (but it can't save it via ASN1). DH on the other hand currently uses a non
+ * FIPS safe prime generator (so it does not have a seed & counter).
+ * The app dhparam also has an option to generate using dsa param generation
+ * which it then converts from DSA to DH.
+ */
+
+#include <string.h>
+#include <openssl/rand.h>
+#include <openssl/sha.h>
+#include <openssl/dsa.h>   /* uses DSS_prime_checks */
+#include "internal/ffc.h"
+
+void FFC_PARAMS_init(FFC_PARAMS *params)
+{
+    memset(params, 0, sizeof(FFC_PARAMS));
+    params->gindex = FFC_UNVERIFIABLE_GINDEX;
+    params->pcounter = -1;
+}
+
+void FFC_PARAMS_cleanup(FFC_PARAMS *params)
+{
+    BN_clear_free(params->p);
+    BN_clear_free(params->q);
+    BN_clear_free(params->g);
+    BN_clear_free(params->j);
+    OPENSSL_clear_free(params->seed, params->seedlen);
+    FFC_PARAMS_init(params);
+}
+
+void FFC_PARAMS_set0_pqg(FFC_PARAMS *d, BIGNUM *p, BIGNUM *q, BIGNUM *g)
+{
+    if (p != NULL && p != d->p) {
+        BN_free(d->p);
+        d->p = p;
+    }
+    if (q != NULL && q != d->q) {
+        BN_free(d->q);
+        d->q = q;
+    }
+    if (g != NULL && g != d->g) {
```

Comment:
I wonder why we do the second check here? It feels like a programming error to pass the same value again since a set0 is supposed to pass ownership of the value - but if it's already set then we shouldn't have attempted to pass ownership again. I mention it only because we don't do this check for `FFC_PARAMS_set0_j` which looks strangely asymmetrical.

Reply:
I didnt want to break how it worked before. Personally I think you should be able to set NULL into the fields if you really want to.